### ISSUE TITLE
CORS microservices : Gateway - UAA - OPTIONS pre flight failure on authentication

### ISSUE THREAD

**Issue by walternate** [jhipster/generator-jhipster#3575]
##### **Overview of the issue**

I know you love these CORS issues so I have decided to add one of my very own. When setting up a gateway and uaa microservice environment, OPTIONS pre-flight requests during authentication are causing CORS issues.

##### **JHipster Version(s)** 

3.2.1

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder** 

Gateway .yo-rc

{
  "generator-jhipster": {
    "jhipsterVersion": "3.2.1",
    "baseName": "gateway",
    "packageName": "com.application.server",
    "packageFolder": "com/application/server",
    "serverPort": "8080",
    "authenticationType": "uaa",
    "uaaBaseName": "uaa",
    "hibernateCache": "no",
    "clusteredHttpSession": "no",
    "websocket": "no",
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "prodDatabaseType": "mongodb",
    "searchEngine": "no",
    "buildTool": "gradle",
    "jwtSecretKey": "c3089601a32274412f570ed2e177fdee1316e9b1",
    "enableTranslation": false,
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "skipClient": true
  }
}

UAA yo-rc

{
  "generator-jhipster": {
    "jhipsterVersion": "3.2.1",
    "baseName": "uaa",
    "packageName": "com.application.server",
    "packageFolder": "com/application/server",
    "serverPort": "9999",
    "authenticationType": "uaa",
    "hibernateCache": "no",
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "prodDatabaseType": "mongodb",
    "searchEngine": "no",
    "buildTool": "gradle",
    "jwtSecretKey": "fbf2b10db6eb8e00737ee95a80175562c8523b07",
    "enableTranslation": false,
    "applicationType": "uaa",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "skipClient": true
  }
}

##### **Browsers and Operating System** 

Mac OSX El Capitan, Chrome 50.0.2661.94

##### **Reference** 

#3318, #3237

##### **Reproduce the error** 

Uncommenting or commenting any combination of the CORS settings in application.yml for both microservices offers no solution and I get:

_XMLHttpRequest cannot load http://localhost:8080/uaa/api/users. The 'Access-Control-Allow-Origin' header contains multiple values 'http://localhost:4200, http://localhost:4200', but only one is allowed. Origin 'http://localhost:4200' is therefore not allowed access._

Here is the cURL:

`curl 'http://localhost:8080/uaa/oauth/token' -X OPTIONS -H 'Access-Control-Request-Method: POST' -H 'Origin: http://localhost:4200' -H 'Accept-Encoding: gzip, deflate, sdch' -H 'Accept-Language: en-US,en;q=0.8' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.94 Safari/537.36' -H 'Accept: */*' -H 'Referer: http://localhost:4200/login' -H 'Connection: keep-alive' -H 'DNT: 1' -H 'Access-Control-Request-Headers: accept, authorization, content-type' --compressed -v`

Addressing the UAA server directly with CORS settings uncommented on the UAA works and I get the 200 response when the POST goes through.

Here is the cURL:

`curl 'http://localhost:9999/oauth/token' -X OPTIONS -H 'Access-Control-Request-Method: POST' -H 'Origin: http://localhost:4200' -H 'Accept-Encoding: gzip, deflate, sdch' -H 'Accept-Language: eser-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.94 Safari/537.36' -H 'Accept: */*' -H 'Referer: http://localhost:4200/login' -H 'Connection: keep-alive' -H 'DNT: 1' -H 'Access-Control-Request-Headers: accept, authorization, content-type' --compressed -v`

The only way I have made it work is by adding the following line in the Gateways WebConfigurer CORSFilter:

`source.registerCorsConfiguration("/*/**", config);`

and uncommenting the Gateways CORS settings and commenting the UAA CORS settings.

It is entirely possible that I don’t know what I am doing and messing up something in the settings but at least it will help others when the solution reveals itself.

**Comment by cbornet** [comment]
Can you try ```source.registerCorsConfiguration("/uaa/oauth/token", config);``` ?

**Comment by walternate** [comment]
Yes, putting `source.registerCorsConfiguration("/uaa/oauth/token", config);` in the gateway CorsFilter does work. I am not sure if there are other cases where OPTIONS will be a problem in other microservices but this works for this specific case. Thanks!

**Comment by deepu105** [comment]
@cbornet do we have to add this to generated code?

**Comment by cbornet** [comment]
I think we do :smile:  !
We should probably also configure the microservice paths on the gateway. Something like:
```
source.registerCorsConfiguration("/*/api/**", config);
```

**Comment by javasparx** [comment]
@cbornet, I also added `source.registerCorsConfiguration("/auth/login/**", config);`. Without it default login process did not work.

```
    @Bean
    public CorsFilter corsFilter() {
        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        CorsConfiguration config = jHipsterProperties.getCors();
        if (config.getAllowedOrigins() != null && !config.getAllowedOrigins().isEmpty()) {
            log.debug("Registering CORS filter");
            source.registerCorsConfiguration("/api/**", config);
            
            //Add this line
            source.registerCorsConfiguration("/auth/login/**", config);
            
            source.registerCorsConfiguration("/management/**", config);
            source.registerCorsConfiguration("/v2/api-docs", config);
            source.registerCorsConfiguration("/*/api/**", config);
            source.registerCorsConfiguration("/*/management/**", config);
            source.registerCorsConfiguration("/*/oauth/**", config);
        }
        return new CorsFilter(source);
    }
```

**Comment by cbornet** [comment]
I don’t know. If the request is blocked by the Cors, then probably. Can you open another issue with all the details?

**Comment by ruddell** [comment]
@javasparx The `/auth` endpoint was added after this issue. I opened a PR with your fix #7552
